frontend/catalyst/qfunc.py

```diff
@@ -106,3 +120,110 @@ def _eval_quantum(*args):
         args_flat = tree_flatten(args)[0]
         res_flat = func_p.bind(flattened_fun, *args_flat, fn=self)
         return tree_unflatten(out_tree_promise(), res_flat)[0]
+
+
+# pylint: disable=protected-access,no-member,not-callable
+def dynamic_one_shot(qnode):
+    """Transform a QNode to into several one-shot tapes to support dynamic circuit execution.
+
+    Args:
+        qnode (QNode): a quantum circuit which will run ``num_shots`` times
+
+    Returns:
+        qnode (QNode):
+
+        The transformed circuit to be run ``num_shots`` times such as to simulate dynamic execution.
+
+
+    **Example**
+
+    Consider the following circuit:
+
+    .. code-block:: python
+
+        dev = qml.device("lightning.qubit", shots=100)
+        params = np.pi / 4 * np.ones(2)
+
+        @qjit
+        @dynamic_one_shot
+        @qml.qnode(dev, diff_method=None)
+        def circuit(x, y):
+            qml.RX(x, wires=0)
+            m0 = measure(0, reset=reset, postselect=postselect)
+
+            @cond(m0 == 1)
+            def ansatz():
+                qml.RY(y, wires=1)
+
+            ansatz()
+            return measure_f(wires=[0, 1])
+
+    The ``dynamic_one_shot`` decorator prompts the QNode to perform a hundred one-shot
+    calculations, where in each calculation the ``measure`` operations dynamically
+    measures the 0-wire and collapse the state vector stochastically.
+    """
+
+    cpy_tape = None
+    aux_tapes = None
+
+    def transform_to_single_shot(qnode):
+        nonlocal cpy_tape
+        nonlocal aux_tapes
```

Comment:
Another question for my understanding, do these declarations do anything given that the variables aren't used in this scope directly, or does one need to declare them nonlocal at each layer of nesting to use them deep within?

Reply:
Seems like they are only required at the innermost level in the present case. There is a nice example, comparing with `global`, [here](https://stackoverflow.com/a/1261961/4150822).